File: backend/internal/domain/card_type.go
package domain

import "strconv"

// CardType identifies a card type line (persisted as smallint on cards.type).
// Values must stay in sync with frontend `src/constants/cardType.js`.
type CardType int16

const (
	CardTypeNonAttackAction CardType = iota
	CardTypeAttackAction
	CardTypeAttackReaction
	CardTypeBlock
	CardTypeCompanion
	CardTypeDefenseReaction
	CardTypeDemiHero
	CardTypeEquipment
	CardTypeHero
	CardTypeInstant
	CardTypeMacro
	CardTypeMentor
	CardTypeResource
	CardTypeToken
	CardTypeWeapon
)

var cardTypeNames = map[CardType]string{
	CardTypeNonAttackAction: "Non-Attack Action",
	CardTypeAttackAction:    "Attack Action",
	CardTypeAttackReaction:  "Attack Reaction",
	CardTypeBlock:           "Block",
	CardTypeCompanion:       "Companion",
	CardTypeDefenseReaction: "Defense Reaction",
	CardTypeDemiHero:        "Demi-Hero",
	CardTypeEquipment:       "Equipment",
	CardTypeHero:            "Hero",
	CardTypeInstant:         "Instant",
	CardTypeMacro:           "Macro",
	CardTypeMentor:          "Mentor",
	CardTypeResource:        "Resource",
	CardTypeToken:           "Token",
	CardTypeWeapon:          "Weapon",
}

// Valid reports whether t is a defined CardType constant.
func (t CardType) Valid() bool {
	_, ok := cardTypeNames[t]
	return ok
}

// String returns the display name or "CardType("+decimal+")" if unknown.
func (t CardType) String() string {
	if name, ok := cardTypeNames[t]; ok {
		return name
	}
	return "CardType(" + strconv.FormatInt(int64(t), 10) + ")"
}

// CardTypes returns every defined type in ascending ID order.
func CardTypes() []CardType {
	return []CardType{
		CardTypeNonAttackAction,
		CardTypeAttackAction,
		CardTypeAttackReaction,
		CardTypeBlock,
		CardTypeCompanion,
		CardTypeDefenseReaction,
		CardTypeDemiHero,
		CardTypeEquipment,
		CardTypeHero,
		CardTypeInstant,
		CardTypeMacro,
		CardTypeMentor,
		CardTypeResource,
		CardTypeToken,
		CardTypeWeapon,
	}
}

File: src/constants/cardType.js
/**
 * Card type line enum IDs (smallint on cards.type). Must match `backend/internal/domain/card_type.go`.
 * @readonly
 */
export const CardType = Object.freeze({
  NonAttackAction: 0,
  AttackAction: 1,
  AttackReaction: 2,
  DefenseReaction: 3,
  Equipment: 4,
  Hero: 5,
  Instant: 6,
  Mentor: 7,
  Resource: 8,
  Token: 9,
  Weapon: 10,
});

/** Ordered display names for each ID 0..10. */
export const CARD_TYPE_NAMES = Object.freeze([
  "Non-Attack Action",
  "Attack Action",
  "Attack Reaction",
  "Defense Reaction",
  "Equipment",
  "Hero",
  "Instant",
  "Mentor",
  "Resource",
  "Token",
  "Weapon",
]);

/**
 * @param {number} id
 * @returns {string | undefined}
 */
export function cardTypeName(id) {
  return CARD_TYPE_NAMES[id];
}

/**
 * @param {number} id
 * @returns {boolean}
 */
export function isValidCardTypeId(id) {
  return Number.isInteger(id) && id >= 0 && id < CARD_TYPE_NAMES.length;
}
